placement/service/placementservice.go

```diff
@@ -219,13 +209,73 @@ func findBestHost(ps placementService, p placement.Snapshot, candidateHosts []pl
 	return nil, errNoValidHost
 }
 
-func (ps placementService) buildRackHostMap(candidateHosts []placement.Host) map[string][]placement.Host {
+func findReplaceHost(s placement.Snapshot, candidateHosts []placement.Host, leaving placement.HostShards, looseRackCheck bool) (placement.Host, error) {
+	// filter out already existing hosts
+	candidateHosts = getNewHostsToPlacement(s, candidateHosts)
+
+	if len(candidateHosts) == 0 {
+		return nil, errNoValidHost
+	}
+	// build rackHostMap from candidate hosts
+	candidateRackHostMap := buildRackHostMap(candidateHosts)
+
+	// if there is a host from the same rack can be added, return it.
+	if hs, exist := candidateRackHostMap[leaving.Host().Rack()]; exist {
+		return hs[0], nil
+	}
+
+	return findHostWithRackConflictCheck(s, candidateRackHostMap, leaving, looseRackCheck)
+}
+
+func findHostWithRackConflictCheck(p placement.Snapshot, rackHostMap map[string][]placement.Host, leaving placement.HostShards, looseRackCheck bool) (placement.Host, error) {
+	// otherwise sort the candidate hosts by the number of conflicts
+	ph := algo.NewPlacementHelper(p)
```

Comment:
If we are constructing a `PlacementHelper` inside the function, maybe this function should just be a method of `PlacementHelper`?